Groovy script to performe clean up on Nexus 3 Docker private repo.

bpankowski opened the pull request.

bpankowski committed:
Groovy script to performe clean up on Nexus 3 Docker private repo.

bpankowski committed:
Added 2nd check on base whitelist with openstack relese.

bpankowski committed:
Added queens to core whitelist and build number 360 to basic wl.

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,117 @@
+        log.info("Processing Component - ${comp.name()}, version: ${comp.version()}");
+        log.info("----------");
+        def fruitEnd = "null";
+        for(i = 0; i < whitelistCore.length; i++){
```
think this loop can be substituted in the following manner

``
 def comp_version = 'rhel-queens-master-395'

 def whitelistCore = ["latest", "ocata", "newton", "queens" ].toArray(); 
 
 def fruitJuice = whitelistCore.any{ item -> item.contains(comp_version)}
 
 print fruitJuice
``

kravvcuCL replied:
Think this loop can be substituted in the following manner

` 
def comp_version = 'rhel-queens-master-395'

def whitelistCore = ["latest", "ocata", "newton", "queens" ].toArray(); 
 
def fruitJuice = whitelistCore.any{ item -> item.contains(comp_version)}
 
return fruitJuice
`

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,117 @@
+def retentionDays = 30;
+def repositoryName = 'BartsDockerRepo';
+def whitelistCore = ["latest", "ocata", "newton", "queens"].toArray();
+def whitelist = ["latest","40", "94", "122", "129", "161", "214", "309", "360"].toArray();
```
These are suffixes, don't see another usage. So rename to e.g. whitelisted_tag_suffixes

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,117 @@
+                def spl = splited.split("-");
+
+                for(j = 0; j < spl.length; j++ ){            
+                    for(k = 0; k < whitelist.length; k++){
```
since whitelist is a list of suffixes you can only check whether the comp.version().endsWith(item)

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,117 @@
+            def fruitJuice = whitelistCore.collect{item -> item.contains(comp.version())}
+            
+            if(fruitJuice[i] == true){
+                fruitEnd = "true";
```
use actual Booleans instead of strings

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,117 @@
+                previousComp = comp.name();
+            }
+        }else{
+            log.info("Component skipped: ${comp.name()}");
```
This line does not make too much sense as it only contains the image name, not the actual tag that was skipped.

bpankowski committed:
Changed structure of checks, abiliti to found out name of builds.

bpankowski committed:
Changed names of check variables.

bpankowski committed:
Change structure of condition for whitelisted tags.

bpankowski committed:
Added ocata and newton to whitelist.

bpankowski committed:
Add log about skipped component on whitelist condition.

bpankowski committed:
Retention counter have been added to the script.

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,97 @@
+        def splited = comp.version();
+        def spl = splited.split("-");
+
+        for (i = spl.length - 1; i < spl.length; i++) {
```
This is a loop over only one element right? Then it's not really a loop :)

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,97 @@
+def retentionDays = 15;
+def retentionBorder = 15;
+def repositoryName = 'BartsDockerRepo';
+def whitelisted_tag_suffixes = ["ocata", "newton", "latest", "40", "94", "122", "129", "161", "214", "309", "360"].toArray();
```
queens missing

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,97 @@
+                Pattern pattern = Pattern.compile(whitelisted_tag_suffixes[j]);
+                Matcher matcher = pattern.matcher(spl[i]);
+                boolean found = matcher.matches();
+                if (found == true) {
```
Correct me if I'm mistaken but looking at the current whitelist this condition may be just an exact match if (spl[i] == whitelisted_tag_suffixes[j])

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,97 @@
+                }
+            }
+            log.info("else check " + prev + " " + spl[i - 1]);
+            if ((spl[i - 1]) == prev) {
```
Dont think this logic with prev makes sense. This would only leave the last 15 tags for e.g. build number 290 so the case matches both queens-master-290 and rhel-queens-master-290 when they should be considered separately.

kravvcuCL replied:
Additionally I think the tx.browseComponents(tx.findBucket(repo)) returns all images and their tags. You need to make sure that each particular image repo is handled separately (that is: leave 15 latest tags and not older than 15 days for contrail-vrouter-agent and contrail-nodemgr etc.). Heres a list of all docker images and their tags on dockerhub: https://hub.docker.com/r/opencontrailnightly/

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,97 @@
+import java.util.regex.Pattern;
+
+def retentionDays = 15;
+def retentionBorder = 15;
```
'threshold' is the word used in IT I think, so rather retentionThreshold

bpankowski committed:
test change

bpankowski committed:
Changed assumption of the problem. Added new cases, fixed check logic.

bpankowski committed:
Delete of not needed for loop.

bpankowski committed:
Delete of not needed parts of code.

Deleted debug logs as well as unwanted parts of code.

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,99 @@
+        def splited = comp.version();
+        def spl = splited.split("-");
+
+            if ((spl[spl.length - 1].isNumber()) == true) {
```
delete

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,99 @@
+    def listOfComponents = ImmutableList.copyOf(components);
+    def previousComp = listOfComponents.head().name();
+
+    listOfComponents.reverseEach { comp ->
```
please add a comment above what we expect the comp.version() to be / it's possible formats

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,99 @@
+                }
+
+            for (j = 0; j < whitelisted_tag_suffixes.length; j++) {
+                if (spl[spl.length - 1] == whitelisted_tag_suffixes[j]) {
```
extract to build_number

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,99 @@
+                    log.info("Component skipped: ${comp.name()} ${comp.version()}");
+                    return checkValue;
+                } else {
+                    if (whitelisted_tag_suffixes.count(spl[spl.length - 1]) == 0) {
```
remove condition

bpankowski replied:
We have to keep that, condition. That's giving us ability to filter out builds thats are not number in that situation e.g. `latest`

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,99 @@
+
+    listOfComponents.reverseEach { comp ->
+        checkValue = null;
+        def splited = comp.version();
```
splited -> tag

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,99 @@
+                    if (whitelisted_tag_suffixes.count(spl[spl.length - 1]) == 0) {
+                        if (tagList.count(spl[spl.length - 1].toInteger()) == 0) {
+                            tagList.add(spl[spl.length - 1].toInteger());
+                            println tagList.sort();
```
asc or desc?

bpankowski replied:
asc

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,99 @@
+        def spl = splited.split("-");
+        def retentionList = tagList.subList(0, tagList.size() - 15);
+
+        if ((spl[spl.length - 1].isNumber()) == true) {
```
remove conditoin

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,99 @@
+        }
+        if (whitelisted_tag_suffixes.count(spl[spl.length - 1]) == 0) {
+            if (retentionList.count(spl[spl.length - 1].toInteger()) > 0) {
+                checkValue = false;
```
unneeded

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,99 @@
+            }
+        }
+        if (whitelisted_tag_suffixes.count(spl[spl.length - 1]) == 0) {
+            if (retentionList.count(spl[spl.length - 1].toInteger()) > 0) {
```
retentionList.contains(build_number)

bpankowski committed:
Cleaning the code. Deleted not usefull code.

Deleted not needed check conditions e.g checkValue. Moved deleting
section to the retention list check part(l. 64-77).

bpankowski committed:
Change names of variables.

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,80 @@
+            if (build_number == whitelisted_tag_suffixes[j]) {
+                log.info("Component skipped: ${comp.name()} ${comp.version()}");
+            } else {
+                if (whitelisted_tag_suffixes.count(build_number) == 0) {
```
IMO this is the same as the 'if' above. Same logic. Actually you do not need the `for` loop if there's a `count` method. This statement here will suffice.

Also I think this can be substituted with the list.contains(item) method.

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,80 @@
+        def tag = comp.version();
+        def tagSplited = tag.split("-");
+        def build_number = tagSplited[tagSplited.length - 1];
+        def retentionList = tagList.subList(0, tagList.size() - 15);
```
extract the magic '15' number to a variable

bpankowski committed:
Add new var with number of lasting builds. Delete of not needed for loop.

bpankowski committed:
Delete of not needed if statment.

bpankowski committed:
Delete of whitespace

bpankowski committed:
Change of if statment.

bpankowski committed:
Bug fixes, Changed retention list and check on date difference.

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -55,11 +55,16 @@ if (components != null) {
             if (retentionList.contains(build_number.toInteger())) {
-                if (comp.lastUpdated() > retentionDate) {
+                if (comp.lastUpdated() < retentionDate) {
                     log.info("retentionDate: ${comp.lastUpdated()} isAfter ${retentionDate}");
```
is older

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -55,11 +55,16 @@ if (components != null) {
+        if(tagList.size() > retentionCount){
+            retentionList = tagList.subList(0, tagList.size() - retentionCount);
+        } else {
+            log.info("Component skipped: ${comp.name()} ${comp.version()}");
```
Change the info to something more meaningful which would say why was the skip done. E.g. 'Number of builds is already less than $retentionCount'

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/clean_up.groovy:
```diff
@@ -0,0 +1,81 @@
+import org.sonatype.nexus.repository.storage.StorageFacet;
```
Renamed the file to docker_cleanup.groovy since we're introducing rpm_cleanup.groovy also.

bpankowski committed:
Renamed file due to new cleaning scripts. Changed logs output.

bpankowski committed:
Add check after branch and build number to the logic.

Whitelist should contain branch name and build number fe. `master-343`
In previous version it wasn't covered.
Add-on instruction on the top of the script.

bpankowski committed:
Add check after branch and build number to the logic.

Whitelist should contain branch name and build number fe. `master-343`
In previous version it wasn't covered.

bpankowski committed:
Update of tags in whitelist.

bpankowski committed:
Update of whitelisted tags. Add of `5.0-168` build.

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/docker_cleanup.groovy:
```diff
@@ -0,0 +1,93 @@
+def retentionCount = 15;
+def tagList = [];
+def repositoryName = 'BartsDockerRepo';
+def whitelisted_tag_suffixes = ["queens", "ocata", "newton", "latest", "5.0-40", "5.0-94", "5.0-122", "5.0-129", "5.0-161", "5.0-168", "5.0-214", "5.0-309", "5.0-360"].toArray();
```
365 is added in case of RPMs, should be added here also?

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/docker_cleanup.groovy:
```diff
@@ -0,0 +1,93 @@
+        def tagSplited = tag.split("-");
+        log.info("tagsplit ${tag} ${tagSplited}");
+        def build_number = tagSplited[tagSplited.length - 1];
+        def two_parts_build_number = tagSplited[tagSplited.length - 2] + "-" + tagSplited[tagSplited.length - 1];
```
Please add some comment on what we try to cover here, an example of a two part build number

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/docker_cleanup.groovy:
```diff
@@ -0,0 +1,93 @@
+        } else if (whitelisted_tag_suffixes.contains(two_parts_build_number)){
+            log.info("Component skipped: ${comp.name()} ${comp.version()}");
+        }else {
+            if (tagList.count(build_number.toInteger()) == 0) {
```
If you want to check whether an object is part of a list use the 'contains' method

https://docs.oracle.com/javase/7/docs/api/java/util/List.html#contains(java.lang.Object)

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/docker_cleanup.groovy:
```diff
@@ -0,0 +1,93 @@
+        }else {
+            if (tagList.count(build_number.toInteger()) == 0) {
+                tagList.add(build_number.toInteger());
+                println tagList.sort();
```
Maybe print after the iteration finishes? outside the 'each' closure? Or is this intended and you want to print the whole list when it grows?

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/docker_cleanup.groovy:
```diff
@@ -0,0 +1,93 @@
+        def build_number = tagSplited[tagSplited.length - 1];
+        def two_parts_build_number = tagSplited[tagSplited.length - 2] + "-" + tagSplited[tagSplited.length - 1];
+        def retentionList = tagList;
+        if(tagList.size() > retentionCount){
```
Think you can move this outside, before the 'each' closure as this if/else should be run only once, to either trim the list or just return

bpankowski committed:
Batch of smaller fixes.

Add of documentation about variables to the script.
Delete of log with tagList used for debug.

bpankowski committed:
Move of if statment outside of each closure.

First check is done on the size of tagList.

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/docker_cleanup.groovy:
```diff
@@ -0,0 +1,94 @@
+    if(tagList.size() > retentionCount){
+        retentionList = tagList.subList(0, tagList.size() - retentionCount);
+        listOfComponents.reverseEach { comp ->
+        def tag = comp.version();
```
Please reformat using indentation. it's hard to distinct code blocks.

kravvcuCL commented on tungsten_ci_utils/nexus_cleanup/docker_cleanup.groovy:
```diff
@@ -0,0 +1,94 @@
+            }
+        }
+    }
+        log.info("\n\nDeleted Component count: ${deletedComponentCount} \n");
```
Please make use of formatting IDE's. This line looks weird between two right parentheses. The parenthesis in the line above matches the one which opens the closure in line 65. And the `listOfComponents` is not in the same column as the parenthesis in line 88.

jabielecki commented:
If someone just fools around and uploads a random tag, the script (1) should eventually delete it (2) shouldn't throw toInteger exception as it does now.

jabielecki commented:
After uncommenting the `delete`, the script only deletes tags but leaves blobs. These can be still downloaded via `docker pull some/thing@sha256:123123`

This work should be postponed until we clarify both the requirements and feasibility.
